### PR TITLE
Forty thieves

### PR DESCRIPTION
**Pull request by keepermanj**
The wiki has more ideas which could be added:

http://keeperrl.com/wiki/index.php?title=Requests_for_new_configurations#The_forty_thieves

### CODE REVIEW

File: level_maker.cpp [hunk 1]
```diff
@@ -2287,6 +2297,9 @@ PLevelMaker LevelMaker::topLevel(RandomGen& random, optional<CreatureFactory> fo
       case SettlementType::CAVE:
         queue = dragonCaveMaker(settlement);
         break;
+      case SettlementType::HUGE_CAVERN:
+        queue = hugeCavernMaker(settlement);
```
Comment (miki151) [inline level_maker.cpp]: The hugeCavenMaker() does something weird, it places stone floors outside, even under trees. Just use dragonCaveMaker() here. We can change the floors later.

Comment (SoftMonster) [inline level_maker.cpp]: There's forty thieves. Will dragonCaveMaker be a bit squashed?

Comment (miki151) [inline level_maker.cpp]: The size will be 35x35 as you defined, it will just use the dragonCaveMaker algorithm. But please verify, I may be wrong.

Comment (keepermanj) [inline level_maker.cpp]: The standard caves don't place any furniture.

Also standard caves always seem to be 12x12 (rather small and a bit repetitive):

Vec2 getSize(RandomGen& random, SettlementType type) {
  switch (type) {
    case SettlementType::WITCH_HOUSE:
    case SettlementType::CEMETERY:
    case SettlementType::MOUNTAIN_LAKE:
    case SettlementType::SMALL_VILLAGE:
    case SettlementType::SWAMP: return {random.get(12, 16), random.get(12, 16)};
    case SettlementType::COTTAGE: return {random.get(8, 10), random.get(8, 10)};
    case SettlementType::FORREST_COTTAGE: return {15, 15};
    case SettlementType::FOREST: return {18, 13};
    case SettlementType::FORREST_VILLAGE: return {20, 20};
    case SettlementType::VILLAGE:
    case SettlementType::ANT_NEST:
    case SettlementType::CASTLE: return {30, 20};
    case SettlementType::CASTLE2: return {13, 13};
    case SettlementType::MINETOWN: return {30, 20};
    case SettlementType::SMALL_MINETOWN: return {15, 15};
    case SettlementType::CAVE: return {12, 12};
    case SettlementType::HUGE_CAVERN: return {35, 35};
    case SettlementType::SPIDER_CAVE: return {12, 12};
    case SettlementType::VAULT: return {10, 10};
    case SettlementType::TOWER: return {5, 5};
    case SettlementType::ISLAND_VAULT_DOOR:
    case SettlementType::ISLAND_VAULT: return {6, 6};
  }
}

Comment (keepermanj) [inline level_maker.cpp]: Would be rather nice to specify the size of the SettlementType rather than it always being fixed.

**Comment by SoftMonster** [comment]
Ah. :) Perhaps we should make the dragonCaveMaker support funiture and even flooring. The flooring I put in for the bandits was buggy though.

**Comment by keepermanj** [comment]
This is actually quite balanced, since the furniture was a bit excessive (there is already loot). Anyway, the flooring would need fixing in "Huge Cavern Maker". The grassy floor is interesting, although I haven't traced down the logic concerning how the floor is actually calculated.

File: creature_factory.cpp [hunk 2]
```diff
@@ -1076,6 +1080,16 @@ CreatureAttributes CreatureFactory::getAttributesFromId(CreatureId id) {
           c.chatReactionHostile = "\"Die!\""_s;
  //         c.skills.insert(SkillId::DISARM_TRAPS);
           c.name = "bandit";);
+    case CreatureId::MASTER_THIEF:
+      //Bigger and badder than bandits.
+      return CATTR(
+          c.viewId = ViewId::MASTER_THIEF;
+          c.attr = LIST(30_dam, 20_def, 100_spd );
+          c.attr[AttrType::SPEED] = 100;
```
Comment (miki151) [inline creature_factory.cpp]: This line is unnecessary.

File: creature_factory.cpp [hunk 3]
```diff
@@ -1281,6 +1295,18 @@ CreatureAttributes CreatureFactory::getAttributesFromId(CreatureId id) {
...
+          c.viewId = ViewId::DARK_KNIGHT;
+          c.attr = LIST(40_dam, 40_def, 110_spd );
+          c.body = Body::humanoid(Body::Size::LARGE);
+          //You really need to defeat this guy with magic.
+          //He is an expert at conventional fighting.
+          c.permanentEffects[LastingEffect::MELEE_RESISTANCE] = 1;
+          c.permanentEffects[LastingEffect::RANGED_RESISTANCE] = 1;
+          c.chatReactionFriendly = "curses all castles"_s;
+          c.chatReactionHostile = "\"Face death!\""_s;
+          c.name = "The Dark Knight";);
```
Comment (miki151) [inline creature_factory.cpp]: Please remove "The", otherwise there will be a double "the" everywhere in the messages.

File: creature_factory.cpp [hunk 4]
```diff
@@ -1076,6 +1080,16 @@ CreatureAttributes CreatureFactory::getAttributesFromId(CreatureId id) {
...
  //         c.skills.insert(SkillId::DISARM_TRAPS);
           c.name = "bandit";);
+    case CreatureId::MASTER_THIEF:
+      //Bigger and badder than bandits.
+      return CATTR(
+          c.viewId = ViewId::MASTER_THIEF;
+          c.attr = LIST(30_dam, 20_def, 100_spd );
+          c.attr[AttrType::SPEED] = 100;
+          c.body = Body::humanoid(Body::Size::LARGE);
+          c.chatReactionFriendly = "Money, money, money..."_s;
```
Comment (miki151) [inline creature_factory.cpp]: I think there need to be extra escaped quotes there.

**Comment by miki151** [comment]
You didn't add buildingId in the enemy definition, that's what causes the grass.

In general I don't like this change, the enemies don't add anything to the game, and there's a LOT of them, which means a huge hack fest to get rid of the villain.

**Comment by keepermanj** [comment]
OK

**Pull request closed without merging**